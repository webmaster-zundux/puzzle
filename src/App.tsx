import "@fontsource-variable/open-sans";
import type { FC } from "react";
import { memo } from "react";
import "./styles.css";

import { PuzzleDataProvider } from "./components/PuzzleDataProvider";
import { PuzzlePage } from "./pages/PuzzlePage";

export const App: FC = memo(function App() {
  return (
    <PuzzleDataProvider>
      <PuzzlePage />
    </PuzzleDataProvider>
  );
});
