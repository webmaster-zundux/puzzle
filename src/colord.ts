export * from "colord";
import * as CD from "colord";
import namesPlugin from "colord/plugins/names";

const { colord: originalColorD, extend } = CD;

extend([namesPlugin]);

export const colord = originalColorD;
