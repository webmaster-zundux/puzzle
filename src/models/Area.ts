import { Point } from "./Point";

export class Area {
  constructor(
    public p0: Point,
    public p1: Point,
    public p2: Point,
    public p3: Point,
  ) {}

  public static create({
    p0 = Point.getZeroPoint(),
    p1 = Point.getZeroPoint(),
    p2 = Point.getZeroPoint(),
    p3 = Point.getZeroPoint(),
  }: {
    p0: Point;
    p1: Point;
    p2: Point;
    p3: Point;
  }) {
    return new Area(p0, p1, p2, p3);
  }

  public static isPointWithinArea(point: Point | { x: number; y: number }, area: Area): boolean {
    const { p0, p2 } = area;
    const { x: px, y: py } = point;

    if (px < p0.x || px > p2.x) {
      return false;
    }

    if (py < p0.y || py > p2.y) {
      return false;
    }

    return true;
  }

  public static isPointWithinAreaByCornerPoints(
    point: Point | { x: number; y: number },
    areaPointP0: Point | { x: number; y: number },
    areaPointP2: Point | { x: number; y: number },
  ): boolean {
    const tx = point.x;
    const ty = point.y;

    const p0 = areaPointP0;
    const p2 = areaPointP2;

    if (tx < p0.x || tx > p2.x) {
      return false;
    }

    if (ty < p0.y || ty > p2.y) {
      return false;
    }

    return true;
  }

  public static doTwoAreasIntersect(areaA: Area, areaB: Area): boolean {
    const aP0 = areaA.p0;
    const aP2 = areaA.p2;

    const bP0 = areaB.p0;
    const bP2 = areaB.p2;

    return Area.doTwoAreasIntersectByCornerPoints(aP0, aP2, bP0, bP2);
  }

  public static doTwoAreasIntersectByCornerPoints(
    areaAPointP0: Point | { x: number; y: number },
    areaAPointP2: Point | { x: number; y: number },
    areaBPointP0: Point | { x: number; y: number },
    areaBPointP2: Point | { x: number; y: number },
  ): boolean {
    const aP0 = areaAPointP0;
    const aP2 = areaAPointP2;

    const bP0 = areaBPointP0;
    const bP2 = areaBPointP2;

    let intersectionOnXAxis = true;
    let intersectionOnYAxis = true;

    if (aP2.x < bP0.x || aP0.x > bP2.x) {
      intersectionOnXAxis = false;
    }

    if (aP2.y < bP0.y || aP0.y > bP2.y) {
      intersectionOnYAxis = false;
    }

    return intersectionOnXAxis && intersectionOnYAxis;
  }

  public static getDefaultArea() {
    return new Area(new Point(0, 0), new Point(1, 0), new Point(1, 1), new Point(0, 1));
  }

  toString() {
    return `Area {p0: ${this.p0}, p1: ${this.p1}, p2: ${this.p2}, p3: ${this.p3}}`;
  }
}
