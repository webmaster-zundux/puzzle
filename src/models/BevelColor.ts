export const BEVEL_EFFECT_HIGHLIGHT = "highlight" as const;
export const BEVEL_EFFECT_SHADOW = "shadow" as const;

export type BEVEL_EFFECT_TYPE = typeof BEVEL_EFFECT_SHADOW | typeof BEVEL_EFFECT_HIGHLIGHT;
