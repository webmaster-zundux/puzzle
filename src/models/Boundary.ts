export class Boundary {
  constructor(
    public x: number,
    public y: number,
    public width: number,
    public height: number,
  ) {}

  public static getIntersectionBoundaryOfTwoBoundaries(a: Boundary, b: Boundary): Boundary | undefined {
    const x1 = a.x > b.x ? a.x : b.x;
    const y1 = a.y > b.y ? a.y : b.y;

    const aP2X = a.x + a.width;
    const aP2Y = a.y + a.height;

    const bP2X = b.x + b.width;
    const bP2Y = b.y + b.height;

    const x2 = aP2X < bP2X ? aP2X : bP2X;
    const y2 = aP2Y < bP2Y ? aP2Y : bP2Y;

    const width = x2 - x1;
    const height = y2 - y1;

    // checking if they are intersect
    if (width <= 0 || height <= 0) {
      return undefined;
    }

    return {
      x: x1,
      y: y1,
      width,
      height,
    };
  }
}
