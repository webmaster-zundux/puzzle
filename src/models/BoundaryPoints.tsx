import type { Point } from "./Point";

export type BoundaryPoints = {
  tl: Point;
  br: Point;
};
