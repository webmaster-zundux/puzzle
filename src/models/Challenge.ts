import type { Id } from "../core/puzzle/BaseEntity";
import { uuid } from "../utils/uuid";

export type ChallengeId = Id;

const generateChallengeIdFunction = () => {
  const newUUID = uuid();
  return newUUID;
};

const EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT =
  import.meta.env.VITE_TESTING_CHALLENGE_ID_GENERATION_EXPOSE_TO_WINDOW_OBJECT === "1";
if (EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT) {
  // @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
  if (typeof window?.generateChallengeId !== "function") {
    Object.defineProperty(window, "generateChallengeId", {
      value: generateChallengeIdFunction,
      writable: false,
      configurable: false,
    });
  }
}

// @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
export const generateChallengeId = window?.generateChallengeId ?? generateChallengeIdFunction;
