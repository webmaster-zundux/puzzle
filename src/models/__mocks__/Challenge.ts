export const generateChallengeId = () => {
  const newUUID = "364bf6d0-6686-4892-be7d-618ac2fcdcc7";
  return newUUID;
};
