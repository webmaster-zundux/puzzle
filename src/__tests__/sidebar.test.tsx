import { act, render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { it, test, vi } from "vitest";

import { App } from "../App";
import { colord } from "../colord";
import * as Challenge from "../models/Challenge";

describe("sidebar", () => {
  describe("puzzle solving challenge timer", () => {
    beforeEach(() => {
      vi.useFakeTimers();
    });

    afterEach(async () => {
      vi.clearAllTimers();
      vi.useRealTimers();
    });

    it("starts when clicking on the canvas", async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
      render(<App />);
      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      const timerTickTimeoutMs = 1000;

      await user.click(screen.getByRole("application", { name: /puzzle view/i }));
      await act(async () => {
        vi.advanceTimersByTime(timerTickTimeoutMs);
      });

      expect(within(sidebar).getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:01");
    });

    it("updates the puzzle elapsed time every second while the puzzle timer is running", async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
      render(<App />);
      const timerTickTimeoutMs = 1000;

      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:00");
      await user.click(screen.getByRole("application", { name: /puzzle view/i }));

      await act(async () => {
        vi.advanceTimersByTime(timerTickTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:01");

      await act(async () => {
        vi.advanceTimersByTime(timerTickTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:02");
    });

    it("pauses when an user idles longer than 60 seconds", async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
      render(<App />);
      const userInactivityTimeoutMs = 60000;

      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:00");
      await user.click(screen.getByRole("application", { name: /puzzle view/i }));

      await act(async () => {
        vi.advanceTimersByTime(userInactivityTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:00");

      await act(async () => {
        const tenSecondsInMs = 10000;
        vi.advanceTimersByTime(tenSecondsInMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:00");
    });

    it("continues to run when an user becomes active after inactivity", async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
      render(<App />);
      const timerTickTimeoutMs = 1000;
      const userInactivityTimeoutMs = 60000;

      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:0:00");
      await user.click(screen.getByRole("application", { name: /puzzle view/i }));
      await act(async () => {
        vi.advanceTimersByTime(userInactivityTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:00");

      await act(async () => {
        const tenSecondsInMs = 10000;
        vi.advanceTimersByTime(tenSecondsInMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:00");

      await user.click(screen.getByRole("application", { name: /puzzle view/i }));
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:00");

      await act(async () => {
        vi.advanceTimersByTime(timerTickTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:01");

      await act(async () => {
        vi.advanceTimersByTime(timerTickTimeoutMs);
      });
      expect(screen.getByRole("note", { name: "elapsed time" })).toHaveTextContent("Time:1:02");
    });

    it.skip("stops when the puzzle challenge solved");
  });

  describe("challenge progress", () => {
    describe("value", () => {
      test.skip(`value increases when two puzzle's pieces connects`);
      test.skip(`value resets to 0% when a new puzzle created`);
      test.skip(`value equals to 100% when the puzzle challenge solved`);
    });

    describe(`button "restart challenge"`, () => {
      it("becomes enabled when the puzzle was touched", async () => {
        render(<App />);
        const sidebar = screen.getByRole("complementary", { name: /sidebar/i });

        const restartPuzzleButton = within(sidebar).getByRole("button", {
          name: /restart puzzle/i,
        });
        expect(restartPuzzleButton).toBeInTheDocument();
        expect(restartPuzzleButton).toBeDisabled();

        await userEvent.click(screen.getByRole("application", { name: /puzzle view/i }));
        await within(sidebar).findByRole("button", { name: /restart puzzle/i });

        await waitFor(() => {
          expect(within(sidebar).getByRole("button", { name: /restart puzzle/i })).toBeEnabled();
        });
      });

      it("becomes disabled when a new puzzle created", async () => {
        vi.spyOn(Challenge, "generateChallengeId")
          .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
          .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");

        render(<App />);

        const restartPuzzleButton = screen.getByRole("button", { name: /restart puzzle/i });
        expect(restartPuzzleButton).toBeInTheDocument();
        expect(restartPuzzleButton).toBeDisabled();

        await userEvent.click(screen.getByRole("application", { name: /puzzle view/i }));
        await waitFor(() => {
          expect(screen.getByRole("note", { name: /challenge seed/i })).toHaveTextContent(
            "b47663e4-95ab-4472-9344-0f165a6a7ff9",
          );
        });

        await waitFor(() => {
          expect(screen.getByRole("button", { name: /restart puzzle/i })).toBeEnabled();
        });

        await userEvent.click(screen.getByRole("button", { name: /restart puzzle/i }));
        await screen.findByRole("dialog", { name: /want to solve the puzzle again\?/i });

        const dialog = screen.getByRole("dialog", {
          name: /want to solve the puzzle again\?/i,
        });
        await userEvent.click(within(dialog).getByRole("button", { name: /solve again/i }));
        await waitFor(() => {
          expect(screen.getByRole("note", { name: /challenge seed/i })).toHaveTextContent(
            "89e80c0c-186a-46db-b242-b147c88a2a3e",
          );
        });

        await waitFor(() => {
          expect(screen.getByRole("button", { name: /restart puzzle/i })).toBeDisabled();
        });
      });
    });
  });

  describe("background color selector", () => {
    it("changes the canvas background color when clicking on any not selected color", async () => {
      const { asFragment } = render(<App />);
      const firstRender = asFragment();

      expect(screen.getByRole("main")).toHaveStyle({
        "background-color": colord("darkslategray").toRgbString(),
      });
      await userEvent.click(screen.getByRole("radio", { name: /white/i }));
      await waitFor(() => {
        expect(screen.getByRole("main")).toHaveStyle({
          "background-color": colord("white").toRgbString(),
        });
      });

      expect(screen.getByRole("radio", { name: /dark slate gray/i })).not.toBeChecked();
      expect(screen.getByRole("radio", { name: /white/i })).toBeChecked();
      expect(firstRender).toMatchDiffSnapshot(asFragment());
    });
  });
});
