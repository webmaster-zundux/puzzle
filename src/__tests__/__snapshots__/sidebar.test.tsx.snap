// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/__tests__/sidebar.test.tsx > sidebar > background color selector > changes the canvas background color when clicking on any not selected color 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,11 @@
    <div
      class="_AppContainer_bf1bbc fullscreen"
    >
      <main
        class="_App_bf1bbc"
-       style="background-color: darkslategray;"
+       style="background-color: white;"
      >
        <div
          class="_CanvasContainer_72d137"
        >
          <canvas
@@ -17,11 +17,11 @@
          />
          <div
            class="_FloatingInfoLabel_72d137"
          >
            <div
-             class="_ZoomLabel_aa5be1"
+             class="_ZoomLabel_aa5be1 _TextForLightBackground_aa5be1"
            >
              <label
                id="zoom-label"
              >
                zoom: "
`;
